Cython/Utility/FunctionArguments.c

```diff
@@ -462,3 +465,352 @@ static CYTHON_INLINE PyObject * __Pyx_GetKwValue_FASTCALL(PyObject *kwnames, PyO
+#define __Pyx_FastcallDict_CLEAR(x, nanny) do { Py_CLEAR(x->object); x->args = NULL; } while(0)
+#define __Pyx_FastcallDict_XDECREF(x, nanny) Py_XDECREF(x->object)
+#define __Pyx_FastcallDict_INCREF __Pyx_FastcallDict_XINCREF
+#define __Pyx_FastcallDict_XINCREF(x, nanny) Py_XINCREF(x->object)
+#define __Pyx_FastcallDict_DECREF_SET __Pyx_FastcallDict_XDECREF_SET
+#define __Pyx_FastcallDict_XDECREF_SET(lhs, rhs) do { \
+        __PyxFastcallDict_obj *temp = lhs; \
+        lhs = rhs; \
+        Py_XDECREF(temp->object); \
+    } while(0)
+#define __Pyx_FastcallDict_NULLCHECK(x) !(x->args!=NULL && x->object==NULL)
+
+/////////////////// FastcallDict //////////////////////
+
+static CYTHON_UNUSED Py_ssize_t __Pyx_FastcallDict_Len(__Pyx_FastcallDict_obj *o) {
+    if (o->object == NULL) {
+        return 0;
+    } else if (o->args) {
+        return PyTuple_GET_SIZE(o->object);
+    } else {
+        return PyDict_Size(o->object);
+    }
+}
+
+static CYTHON_UNUSED __Pyx_FastcallDict_obj __Pyx_FastcallDict_New(void) {
+    __Pyx_FastcallDict_obj out = {};
+    return out;
+}
+
+/////////////////// FastcallDictConvert.proto //////////////////////
+
+static CYTHON_UNUSED PyObject* __Pyx_FastcallDict_ToDict(__Pyx_FastcallDict_obj *o); /* proto */
+static CYTHON_UNUSED PyObject* __Pyx_FastcallDict_ToDict_Explicit(__Pyx_FastcallDict_obj *o); /* proto */
+
+/////////////////// FastcallDictConvert //////////////////////
+//@requires: fastcall
+
+static CYTHON_UNUSED PyObject* __Pyx_FastcallDict_ToDict(__Pyx_FastcallDict_obj *o){
+    // when implicitly converted to dict change o.object so that modifications
+    // to the dict propagate
+    PyObject* dict = __Pyx_FastcallDict_ToDict_Explicit(o);
+    if (!dict) return NULL;
+    o->args = NULL;
+    Py_CLEAR(o->object);
+    o->object = dict;
+    Py_INCREF(dict);
+    return dict;
+}
+static CYTHON_UNUSED PyObject* __Pyx_FastcallDict_ToDict_Explicit(__Pyx_FastcallDict_obj* o) {
+    if (o->object == NULL) {
+        return PyDict_New();
+    } else if (o->args) {
+        return __Pyx_KwargsAsDict_FASTCALL(o->object, o->args);
+    } else { // already a dict
+        Py_INCREF(o->object);
+        return o->object;
+    }
+}
+
+/////////////////// FastcallDictIter.proto //////////////////////
+
+// These methods behaves slightly differently from the regular dict methods
+// in that they just return a list rather than an iterator. For most of them
+// an iterator would be dangerous since it might outlive the __Pyx_FastcallDict_obj
+static CYTHON_UNUSED PyObject* __Pyx_FastcallDict_Keys(__Pyx_FastcallDict_obj* o); /* proto */
+static CYTHON_UNUSED PyObject* __Pyx_FastcallDict_Values(__Pyx_FastcallDict_obj* o); /* proto */
+// Items is excluded since it's more complicated and performance wasn't great
+
+/////////////////// FastcallDictIter //////////////////////
+//@requires:ObjectHandling.c::TupleAndListFromArray
+
+static CYTHON_UNUSED PyObject* __Pyx_FastcallDict_Keys(__Pyx_FastcallDict_obj *o) {
```

Comment:
We need to return the expected type here, both in Py2 (list) and Py3 (view).

Reply:
This suggests to me that both `keys` and `values` should probably just lead to a coercion to a Python dict instead. Anything that emulates a view is going to need a lot of implementing.